### ISSUE TITLE
unexpected result when arranging after grouping by multiple variables

### ISSUE THREAD

**Issue by justmarkham** [hadley/dplyr#986]
In this code, I'm grouping mtcars by cylinder, counting the number of rows in each group, and arranging the results in descending order by the count:

```
mtcars %>% group_by(cyl) %>% summarise(cnt=n()) %>% arrange(desc(cnt))
```

I get the expected result:

```
  cyl cnt
1   8  14
2   4  11
3   6   7
```

Now I'll group by cylinder and gear:

```
mtcars %>% group_by(cyl, gear) %>% summarise(cnt=n()) %>% arrange(desc(cnt))
```

This time, I get a result I'm not expecting:

```
  cyl gear cnt
1   4    4   8
2   4    5   2
3   4    3   1
4   6    4   4
5   6    3   2
6   6    5   1
7   8    3  12
8   8    5   2
```

I'm expecting the entire dataframe to sort in descending order by "cnt" (thus 8 3 12 would be the first row, 4 4 8 would be the second row, etc). Instead, it is sorting by "cyl", and then within each "cyl" group, it is sorting in descending order by "cnt".

Is this a bug, or am I misunderstanding how this is supposed to work?

Side note: using the tally function also produces the same results as above:

```
mtcars %>% group_by(cyl) %>% tally(sort=TRUE)

  cyl  n
1   8 14
2   4 11
3   6  7

mtcars %>% group_by(cyl, gear) %>% tally(sort=TRUE)

  cyl gear  n
1   4    4  8
2   4    5  2
3   4    3  1
4   6    4  4
5   6    3  2
6   6    5  1
7   8    3 12
8   8    5  2
```

I'm using dplyr 0.4.1. For what it's worth, this used "work" back in version 0.2 (meaning, it used to produce the results I was expecting).

#491 (which was resolved in 0.3) might be related to this issue.

**Comment by DavidArenburg** [comment]
`dplyr` drops one grouping level after each aggregation operation by design.

**Comment by justmarkham** [comment]
Thanks for the information, @DavidArenburg. Do you know of a straightforward way in dplyr to achieve the results I was expecting?

**Comment by DavidArenburg** [comment]
```r
mtcars %>% group_by(cyl, gear) %>% summarise(cnt=n()) %>% arrange(gear, desc(cnt))
```
Maybe? Thugh there could be probably a smarter way

**Comment by justmarkham** [comment]
@DavidArenburg Thanks for the suggestion, but unfortunately that does not achieve the desired result. Here's what it outputs:

```
  cyl gear cnt
1   4    3   1
2   4    4   8
3   4    5   2
4   6    3   2
5   6    4   4
6   6    5   1
7   8    3  12
8   8    5   2
```

**Comment by justmarkham** [comment]
Ah! I figured out how to produce the result I was expecting:

```
mtcars %>% group_by(cyl, gear) %>% summarise(cnt=n()) %>% ungroup() %>% arrange(desc(cnt))

  cyl gear cnt
1   8    3  12
2   4    4   8
3   6    4   4
4   4    5   2
5   6    3   2
6   8    5   2
7   4    3   1
8   6    5   1
```

I'm fine with having this issue closed, since the documentation appears to support the idea that the behavior I reported is the "expected" behavior, and is not a bug. However, I'd prefer for someone who actually works on dplyr to close it, since they would be in a better position than me to confirm that this is not a bug.

Thanks!
